ccan/ccan/graphql/graphql.h

```diff
@@ -0,0 +1,230 @@
+	struct graphql_directives *directives;
+	struct graphql_selection_set *sel_set;
+};
+
+struct graphql_type_condition {
+	struct graphql_named_type *named_type;
+};
+
+struct graphql_fragment_name {
+	struct graphql_token *name;
+};
+
+struct graphql_fragment_definition {
+	struct graphql_fragment_name *name;
+	struct graphql_type_condition *type_cond;
+	struct graphql_directives *directives;
+	struct graphql_selection_set *sel_set;
+};
+
+struct graphql_fragment_spread {
+	struct graphql_fragment_name *name;
+	struct graphql_directives *directives;
+};
+
+struct graphql_alias {
+	struct graphql_token *name;
+};
+
+struct graphql_argument {
+	struct graphql_argument *next;
+	struct graphql_token *name;
+	struct graphql_value *val;
+};
+
+struct graphql_arguments {
+	struct graphql_argument *first;
+};
+
+struct graphql_field {
+	struct graphql_alias *alias;
+	struct graphql_token *name;
+	struct graphql_arguments *args;
+	struct graphql_directives *directives;
+	struct graphql_selection_set *sel_set;
+};
+
+struct graphql_selection {
+	struct graphql_selection *next;
+	struct graphql_field *field;
+	struct graphql_fragment_spread *frag_spread;
+	struct graphql_inline_fragment *inline_frag;
+};
+
+struct graphql_selection_set {
+	struct graphql_selection *first;
+};
+
+struct graphql_operation_type {
+	struct graphql_token *op_type;
+};
+
+struct graphql_operation_definition {
+	struct graphql_operation_type *op_type;
+	struct graphql_token *op_name;
+	struct graphql_variable_definitions *vars;
+	struct graphql_directives *directives;
+	struct graphql_selection_set *sel_set;
+};
+
+struct graphql_executable_definition {
+	struct graphql_executable_definition *next_def;
+	struct graphql_operation_definition *op_def;
+	struct graphql_fragment_definition *frag_def;
+};
+
+struct graphql_executable_document {
+	struct graphql_executable_definition *first_def;
+};
+
+struct graphql_definition {
+	struct graphql_definition *next_def;
+	struct graphql_executable_definition *executable_def;
+	struct graphql_type_system_definition_or_extension *type_system_def;
+};
+
+struct graphql_document {
+	struct graphql_definition *first_def;
+};
+
+
+struct graphql_token {
+	struct list_node list;
+	unsigned int token_type;
+	unsigned int token_specific;
+	char *token_string;
+	unsigned int source_line;
+	unsigned int source_column;
+	unsigned int source_len;
+};
+
+/* The lexer.
+ * INPUTS:
+ *	input - string to parse
+ *	ctx - parent tal context or NULL
+ *	tokens - a variable to receive the resulting token list
+ * RETURN:
+ *	GRAPHQL_SUCCESS or an error string.
+ */
+const char *graphql_lex(const char *input, const tal_t *ctx, struct list_head **tokens);
+
+/* The parser.
+ * INPUTS:
+ *	tokens - the list produced by the lexer
+ *	doc - a variable to receive the resulting abstract syntax tree (AST)
+ * RETURN:
+ *	GRAPHQL_SUCCESS or an error string.
+ */
+const char *graphql_parse(struct list_head *tokens, struct graphql_executable_document **doc);
```

Comment:
You should probably document that the tokens list is emptied by this function?

Reply:
Good idea, done.